---
title: "Fetal Tissue Gene Expression EDA"
author: "Mike Lafferty"
date: "9/11/2018"
output: 
  html_document:
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(here)
library(dplyr)
library(readr)
library(magrittr)
library(DESeq2)
library(ggplot2)
library(GGally)
library(limma)
library(reshape2)

source(here("src/utils/lafferty_utils.R"))
```

## Build DDS

```{r}
# INPUT FILES #####################################################################################
# ranged summarized experiment with gene expression counts
expression_rse_rds <- here("results/rdata_files/20181018_fetalTissue_ranged_summarized_experiment_gene_counts.rds")
# summarized experiment with mirge2.0 quantified mirna counts from mirbase v22
mirbase_se_rds <- here("results/rdata_files/20180906_mirbase_v22_mirge2.0_counts.rds")

# Build DDS miRNA data #######################################################################################
# read in summarized experiment
se <- readRDS(mirbase_se_rds)

# label outliers by PCA (after batch effect removal of pool and rna purif. method, see below)
outlier_rnaid <- c("RNAID1710", "RNAID1744", "RNAID1698", "RNAID1814", "RNAID1678")
colData(se)$outlier <- ifelse(colData(se)$rnaid %in% outlier_rnaid, "yes", "no")

# build DESeq data set
dds_mirna <- DESeqDataSet(se, design = ~1)
# remove rows with only zero counts
dds_mirna <- dds_mirna[rowSums(counts(dds_mirna)) > 1, ]
# estimate size factors for normalization
dds_mirna <- estimateSizeFactors(dds_mirna)
# norm. transformed counts (DESeq size factors, log2 transformed)
#ntc <- assay(normTransform(dds_mirna))
# variance stabilizing tranformed data
vsd_mirna <- varianceStabilizingTransformation(dds_mirna)
rm(se)

# Build DDS gene expression data #######################################################################################
# read in ranged summarized experiment
rse <- readRDS(expression_rse_rds)

# label outliers by PCA (after batch effect removal of pool and rna purif. method, see below)
#outlier_rnaid <- c("RNAID1710", "RNAID1744", "RNAID1698", "RNAID1814", "RNAID1678")
#colData(se)$outlier <- ifelse(colData(se)$rnaid %in% outlier_rnaid, "yes", "no")

# build DESeq data set
dds <- DESeqDataSet(rse, design = ~1)
# remove rows with only zero counts
dds <- dds[rowSums(counts(dds)) > 1, ]
# filter for samples within mirna dataset
dds <- dds[,which(colnames(dds) %in% dds_mirna$rnaid)]
# estimate size factors for normalization
dds <- estimateSizeFactors(dds)
# norm. transformed counts (DESeq size factors, log2 transformed)
#ntc <- assay(normTransform(dds))
# variance stabilizing tranformed data
vsd <- vst(dds)
rm(rse)
```

## Sequencing and Library Depth

Incomplete metadata for total RNA-Seq runs.

```{r}
#df <- as.data.frame(colData(dds))


```

## PCA

Center=TRUE, Scale=FALSE

```{r}
pca <- prcomp(t(assay(vsd)), center = TRUE, scale. = FALSE)
df <- data.frame(pca$x[,1:5], colData(vsd))
percentVar <- pca$sdev^2 / sum(pca$sdev^2) * 100
```

### Scree Plot

```{r}
ggplot(data.frame(perc_var=percentVar, pc=1:length(percentVar))[1:10,], aes(x=pc, y=perc_var)) +
  geom_point() +
  geom_line() +
  labs(x="PC", y="Percent Total Variance",
       title="PCA: Gene Expression (VST Norm.)") +
  plotTheme
```

### PC1 v PC2

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=tissue_section)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=c("blue", "grey50", "darkorange")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="center = TRUE, scale. = FALSE")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=gestation_week)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_gradientn(colors = c("darkorange", "grey80", "blue")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="center = TRUE, scale. = FALSE")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=rin)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_gradientn(colors = c("darkorange", "grey80", "blue")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="center = TRUE, scale. = FALSE")
```



```{r}
df %>%
  ggplot(aes(PC1, PC2, color=round)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=c("blue", "grey50", "darkorange")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="center = TRUE, scale. = FALSE")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=rna_purification_method)) +
  geom_point() +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=cbPalette) +
  labs(title="Gene Expression (VST Norm.)",
       caption="center = TRUE, scale. = FALSE")
```

## Xist Expression

```{r}
df <- data.frame(df,
                 xist_exp_vst = assay(vsd)["ENSG00000229807",],
                 xist_scaled_cts = counts(dds, normalized=TRUE)["ENSG00000229807",])

df %>%
  ggplot(aes(y=log2(xist_scaled_cts), x=sex)) +
  geom_point() +
  plotTheme

df %>%
  dplyr::select(rnaid, sex, xist_scaled_cts) -> tmp
```

## Remove Batch Effects

### Round Only

```{r}
vsd.trans <- vsd
assay(vsd.trans) <- limma::removeBatchEffect(assay(vsd),
                                             batch = vsd$round)
```

#### PCA

```{r}
pca <- prcomp(t(assay(vsd.trans)), center = TRUE, scale. = FALSE)
df <- data.frame(pca$x[,1:5], colData(vsd.trans))
percentVar <- pca$sdev^2 / sum(pca$sdev^2) * 100
```

```{r}
ggplot(data.frame(perc_var=percentVar, pc=1:length(percentVar))[1:10,], aes(x=pc, y=perc_var)) +
  geom_point() +
  geom_line() +
  labs(x="PC", y="Percent Total Variance",
       title="PCA: Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round") +
  plotTheme
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=tissue_section)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=c("blue", "grey50", "darkorange")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=gestation_week)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_gradientn(colors = c("darkorange", "grey80", "blue")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=rin)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_gradientn(colors = c("darkorange", "grey80", "blue")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round")
```



```{r}
df %>%
  ggplot(aes(PC1, PC2, color=round)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=c("blue", "grey50", "darkorange")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=rna_purification_method)) +
  geom_point() +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=cbPalette) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round")
```

### Round and RIN

```{r}
vsd.trans <- vsd
assay(vsd.trans) <- limma::removeBatchEffect(assay(vsd),
                                             batch = vsd$round,
                                             batch2 = vsd$rin)
```

#### PCA

```{r}
pca <- prcomp(t(assay(vsd.trans)), center = TRUE, scale. = FALSE)
df <- data.frame(pca$x[,1:5], colData(vsd.trans))
percentVar <- pca$sdev^2 / sum(pca$sdev^2) * 100
```

```{r}
ggplot(data.frame(perc_var=percentVar, pc=1:length(percentVar))[1:10,], aes(x=pc, y=perc_var)) +
  geom_point() +
  geom_line() +
  labs(x="PC", y="Percent Total Variance",
       title="PCA: Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin") +
  plotTheme
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=tissue_section)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=c("blue", "grey50", "darkorange")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=gestation_week)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_gradientn(colors = c("darkorange", "grey80", "blue")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=rin)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_gradientn(colors = c("darkorange", "grey80", "blue")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin")
```



```{r}
df %>%
  ggplot(aes(PC1, PC2, color=round)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=c("blue", "grey50", "darkorange")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=rna_purification_method)) +
  geom_point() +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=cbPalette) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin")
```

### Round and RIN Preserving Gest. Week

```{r}
vsd.trans <- vsd
assay(vsd.trans) <- limma::removeBatchEffect(assay(vsd),
                                             batch = vsd$round,
                                             batch2 = vsd$rin,
                                             design = model.matrix(~vsd$gestation_week))
```

#### PCA

```{r}
pca <- prcomp(t(assay(vsd.trans)), center = TRUE, scale. = FALSE)
df <- data.frame(pca$x[,1:5], colData(vsd.trans))
percentVar <- pca$sdev^2 / sum(pca$sdev^2) * 100
```

```{r}
ggplot(data.frame(perc_var=percentVar, pc=1:length(percentVar))[1:10,], aes(x=pc, y=perc_var)) +
  geom_point() +
  geom_line() +
  labs(x="PC", y="Percent Total Variance",
       title="PCA: Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin, pres. gestation week") +
  plotTheme
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=tissue_section)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=c("blue", "grey50", "darkorange")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin, pres. gestation week")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=gestation_week)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_gradientn(colors = c("darkorange", "grey80", "blue")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin, pres. gestation week")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=rin)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_gradientn(colors = c("darkorange", "grey80", "blue")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin, pres. gestation week")
```



```{r}
df %>%
  ggplot(aes(PC1, PC2, color=round)) +
  geom_point(size=2) +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=c("blue", "grey50", "darkorange")) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin, pres. gestation week")
```

```{r}
df %>%
  ggplot(aes(PC1, PC2, color=rna_purification_method)) +
  geom_point() +
  labs(x=paste("PC1 (", round(percentVar[1], 1), "%)", sep=""),
       y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
  plotTheme +
  scale_color_manual(values=cbPalette) +
  labs(title="Gene Expression (VST Norm.)",
       caption="removeBatchEffect: round and rin, pres. gestation week")
```

#### All Tech. Covariates

##### PC1

```{r}
for (variable in colnames(df)) {
  df %>%
    ggplot(aes_string(x=variable, y="PC1")) +
    geom_point(size=2) +
    labs(y=paste("PC1 (", round(percentVar[1], 1), "%)", sep="")) +
    plotTheme +
    labs(title="PCA: Gene Expression") -> p
    
    print(p)
}
```

##### PC2

```{r}
for (variable in colnames(df)) {
  df %>%
    ggplot(aes_string(x=variable, y="PC2")) +
    geom_point(size=2) +
    labs(y=paste("PC2 (", round(percentVar[2], 1), "%)", sep="")) +
    plotTheme +
    labs(title="PCA: Gene Expression") -> p
    
    print(p)
}
```

##### PC3

```{r}
for (variable in colnames(df)) {
  df %>%
    ggplot(aes_string(x=variable, y="PC3")) +
    geom_point(size=2) +
    labs(y=paste("PC3 (", round(percentVar[3], 1), "%)", sep="")) +
    plotTheme +
    labs(title="PCA: Gene Expression") -> p
    
    print(p)
}
```

##### PC4

```{r}
for (variable in colnames(df)) {
  df %>%
    ggplot(aes_string(x=variable, y="PC4")) +
    geom_point(size=2) +
    labs(y=paste("PC4 (", round(percentVar[4], 1), "%)", sep="")) +
    plotTheme +
    labs(title="PCA: Gene Expression") -> p
    
    print(p)
}
```

##### PC5

```{r}
for (variable in colnames(df)) {
  df %>%
    ggplot(aes_string(x=variable, y="PC5")) +
    geom_point(size=2) +
    labs(y=paste("PC5 (", round(percentVar[5], 1), "%)", sep="")) +
    plotTheme +
    labs(title="PCA: Gene Expression") -> p
    
    print(p)
}
```
