---
title: "HNP Differentiation Microarray Differential Expression: Relationships"
author: "Mike Lafferty"
date: "2022-09-21"
output: 
  html_document:
    toc: yes
    toc_depth: 5
    toc_float: true
    code_folding: hide
---

```{r setup, include=TRUE, echo=TRUE, message=FALSE, class.source = "fold-show", warning=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE)
library(here)
library(readr)
library(dplyr)
library(tidyr)
library(readxl)
library(magrittr)
library(ggplot2)
library(ggrepel)

library(Biobase)
library(pd.clariom.s.human.ht)
library(clariomshumantranscriptcluster.db)
library(oligo)

library(limma)

library(topGO)
library(org.Hs.eg.db)

library(multiMiR)
library(miRNAtap)

library(ComplexHeatmap)

library(mikelaffr)
library(knitr)

# OUTPUT #############################################################
dir.pdfs <- here("doc/microarray/pdfs/")
dir.create(dir.pdfs, recursive = TRUE, showWarnings = FALSE)

dir.topgo <- here("doc/microarray/topgo/")
dir.create(dir.topgo, recursive = TRUE, showWarnings = FALSE)
# INPUT ##############################################################
# Microarray ExpressionSet for HNP Differentiataion Experiment, normalized, probe filtered
expressionSet.rds <- here("results/rdata_files/20220829_es_HNP_Differential_Microarray_SST-RMA.rds")

# single cell data from luis at UCLA showing genes upregulated in cell type clusters
#genes.by.cluster.xlsx <- here("data/ucla_single_cell/TableS4 Cluster analysis.xlsx")

# TargetScan 8.0 Predicted Targets for miR-4707-3p
targetScan8.0.mir4707.targets.txt <- here("data/target_predictions/TargetScan8.0_miR-4707-3p.predicted_targets.txt")

# miRDB v6.0 Target Predictions
miRDB6.0.target.predictions.txt <- here("data/target_predictions/miRDB_v6.0_prediction_result.txt.gz")

# miRTarBase v7 Validated Target Interactions
miRTarBase7.target.interactions.xlsx <- here("data/mirtarbase_v7/hsa_MTI.xlsx")

# GLOBALS ############################################################
```

## Differential Expression

```{r, class.source = "fold-show"}
# Import ExpressionSet
eSet <- readRDS(expressionSet.rds)

# add condition label
eSet$Condition <- paste(eSet$Timepoint, eSet$Expression, sep = "_")

# remove outliers
filt.eSet <- eSet[,!eSet$outlier]

# corrected expression for plotting
corrected.expression <- limma::removeBatchEffect(filt.eSet,
                                                 covariates = filt.eSet$cDNA_Yield_ug,
                                                 design = model.matrix(~filt.eSet$Timepoint + filt.eSet$Expression))

# conduct differential expression primarily based on the 4 condition groups
condition <- factor(filt.eSet$Condition)

# include cDNA_yield in the model matrix
design <- model.matrix(~ 0 + filt.eSet$cDNA_Yield_ug + condition)
colnames(design) <- c("cDNA_yield", levels(condition))

# fit to the expression data
fit <- limma::lmFit(filt.eSet, design)

# contrasts based on expression (contrl v 4707) or differentiation (week1 v week2)
cont.matrix <- makeContrasts(Expression_in_Week1 = Week1_Control-Week1_4707,
                             Expression_in_Week2 = Week2_Control-Week2_4707,
                             Diff_in_Control = Week2_Control-Week1_Control,
                             Diff_in_4707 = Week2_4707-Week1_4707,
                             Interaction = (Week2_Control-Week2_4707)-(Week1_Control-Week1_4707),
                             levels = design)

# fit for contrasts
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2)

# results tables
df.results.expression_in_week1 <- as_tibble(topTable(fit2, coef = "Expression_in_Week1", number=Inf, adjust.method = "BH"))
df.results.expression_in_week2 <- as_tibble(topTable(fit2, coef = "Expression_in_Week2", number=Inf, adjust.method = "BH"))
df.results.diff_in_control <- as_tibble(topTable(fit2, coef = "Diff_in_Control", number=Inf, adjust.method = "BH"))
df.results.diff_in_4707 <- as_tibble(topTable(fit2, coef = "Diff_in_4707", number=Inf, adjust.method = "BH"))
df.results.interaction <- as_tibble(topTable(fit2, coef = "Interaction", number=Inf, adjust.method = "BH"))


rm(condition, design, fit, cont.matrix)
```

**Differentiation:**  
__+LFC__: upregulated in week 2 (neuron markers?)  
__-LFC__: upregulated in week 1 (progenitor markers?)  

**Expression:**  
__+LFC__: upregulated in CONTROL (repressed by mir4707?)  
__-LFC__: upregulated in mir4707 induced  


## miR-4707-3p Targets

```{r, class.source = "fold-show", message=FALSE}
df.targetScan.predictions <- read_tsv(targetScan8.0.mir4707.targets.txt)

df.mirdb.predictions <- read_tsv(miRDB6.0.target.predictions.txt, col_names = c("miRNA_Name", "GenBank", "Target_Score", "ENTREZID"))
df.mirdb.predictions %<>%
    filter(miRNA_Name == "hsa-miR-4707-3p") %>%
    filter(!duplicated(ENTREZID)) %>%
    mutate(ENTREZID = as.character(ENTREZID))

anno <- AnnotationDbi::select(org.Hs.eg.db,
                              keys = df.mirdb.predictions$ENTREZID,
                              keytype = "ENTREZID",
                              columns = c("SYMBOL"))

df.mirdb.predictions %<>%
    left_join(anno, by = "ENTREZID")

df.mirtarbase.interactions <- read_xlsx(miRTarBase7.target.interactions.xlsx)
df.mirtarbase.interactions %<>%
    filter(miRNA == "hsa-miR-4707-3p") %>%
    filter(!duplicated(`Target Gene`))

rm(anno)
```

**miRDB Predicted Targets**: 46  

**TargetScan Predicted Targets**: 1758  

**miRTarBase Validated Interactions**: 104  


```{r, class.source = "fold-show"}
table(df.mirdb.predictions$SYMBOL %in% df.targetScan.predictions$`Target gene`)

table(df.mirdb.predictions$SYMBOL %in% df.mirtarbase.interactions$`Target Gene`)

table(df.mirtarbase.interactions$`Target Gene` %in% df.targetScan.predictions$`Target gene`)
```

## Expression Categories

```{r}
# all expressed genes
all.genes <- fData(filt.eSet)$SYMBOL

# combine expression data into one dataframe

df.results.diff_in_control %<>%
    mutate(Sig_at_0.05 = adj.P.Val < 0.05,
           Sig_at_0.1 = adj.P.Val < 0.1)

df.results.diff_in_4707 %<>%
    mutate(Sig_at_0.05 = adj.P.Val < 0.05,
           Sig_at_0.1 = adj.P.Val < 0.1)

df.results.expression_in_week1 %<>%
    mutate(Sig_at_0.05 = adj.P.Val < 0.05,
           Sig_at_0.1 = adj.P.Val < 0.1)

df.results.expression_in_week2 %<>%
    mutate(Sig_at_0.05 = adj.P.Val < 0.05,
           Sig_at_0.1 = adj.P.Val < 0.1)

df.results.interaction %<>%
    mutate(Sig_at_0.05 = adj.P.Val < 0.05,
           Sig_at_0.1 = adj.P.Val < 0.1)

df.results.diff <- left_join(df.results.diff_in_control, df.results.diff_in_4707,
                             by = c("PROBEID", "SYMBOL", "GENENAME", "ENSEMBL", "ENTREZID", "duplicate.annotation"),
                             suffix = c(".diff_in_control", ".diff_in_4707"))


df.results.expression <- left_join(df.results.expression_in_week1, df.results.expression_in_week2,
                                   by = c("PROBEID", "SYMBOL", "GENENAME", "ENSEMBL", "ENTREZID", "duplicate.annotation"),
                                   suffix = c(".expression_in_week1", ".expression_in_week2"))

df.results <- left_join(df.results.diff, df.results.expression,
                        by = c("PROBEID", "SYMBOL", "GENENAME", "ENSEMBL", "ENTREZID", "duplicate.annotation"))

df.results.interaction %>%
    dplyr::select(PROBEID,
           logFC.interaction = logFC,
           AveExpr.interaction = AveExpr,
           t.interaction = t,
           P.Value.interaction = P.Value,
           adj.P.Val.interaction = adj.P.Val,
           B.interaction = B,
           Sig_at_0.05.interaction = Sig_at_0.05,
           Sig_at_0.1.interaction = Sig_at_0.1) -> df.results.int

df.results %<>%
    left_join(df.results.int, by = "PROBEID")

df.results %<>%
    mutate(miRDB_predicted_target = SYMBOL %in% df.mirdb.predictions$SYMBOL,
           TargetScan_predicted_target = SYMBOL %in% df.targetScan.predictions$`Target gene`,
           miRTarBase_validated_target = SYMBOL %in% df.mirtarbase.interactions$`Target Gene`)

rm(df.results.int)
```


```{r}
# categorize each gene based on the 4 differential expression relationships, using "flags" at each category
# Flag1: Differentiation DE (week1 v week2) in CONTROL
# Flag2: Differentiation DE (week1 v week2) in 4707
# Flag3: Expression DE (Ctrl v 4707) in Week1
# Flag4: Expression DE (Ctrl v 4707) in Week2

# Flags based on direction of logFC (LFC) and whether there is a significant difference
# -: negative LFC
# 0: non-significant LFC
# +: positive LFC

# example flag: -0+0
# negative LFC for differentiation in CTRL, non-sig LFC for diff in 4707, positive lFC for expression in week1, non-sig LFC for expression in week2


df.results %<>%
    mutate(DE_flag_at_0.05 = NA,
           DE_flag_at_0.1 = NA)

# loop over each gene and calculate flag
for (i in 1:nrow(df.results)) {
    flag_at_0.05 <- ""
    flag_at_0.1 <- ""
    
    # Flag1: Differentiation DE (week1 v week2) in CONTROL
    if (!df.results$Sig_at_0.05.diff_in_control[i]) {
        flag_at_0.05 <- paste0(flag_at_0.05, "0")
    } else if (df.results$logFC.diff_in_control[i] < 0) {
        flag_at_0.05 <- paste0(flag_at_0.05, "-")
    } else if (df.results$logFC.diff_in_control[i] > 0) {
        flag_at_0.05 <- paste0(flag_at_0.05, "+")
    }
    
    if (!df.results$Sig_at_0.1.diff_in_control[i]) {
        flag_at_0.1 <- paste0(flag_at_0.1, "0")
    } else if (df.results$logFC.diff_in_control[i] < 0) {
        flag_at_0.1 <- paste0(flag_at_0.1, "-")
    } else if (df.results$logFC.diff_in_control[i] > 0) {
        flag_at_0.1 <- paste0(flag_at_0.1, "+")
    }
    
    # Flag2: Differentiation DE (week1 v week2) in 4707
    if (!df.results$Sig_at_0.05.diff_in_4707[i]) {
        flag_at_0.05 <- paste0(flag_at_0.05, "0")
    } else if (df.results$logFC.diff_in_4707[i] < 0) {
        flag_at_0.05 <- paste0(flag_at_0.05, "-")
    } else if (df.results$logFC.diff_in_4707[i] > 0) {
        flag_at_0.05 <- paste0(flag_at_0.05, "+")
    }
    
    if (!df.results$Sig_at_0.1.diff_in_4707[i]) {
        flag_at_0.1 <- paste0(flag_at_0.1, "0")
    } else if (df.results$logFC.diff_in_4707[i] < 0) {
        flag_at_0.1 <- paste0(flag_at_0.1, "-")
    } else if (df.results$logFC.diff_in_4707[i] > 0) {
        flag_at_0.1 <- paste0(flag_at_0.1, "+")
    }
    
    # Flag3: Expression DE (Ctrl v 4707) in Week1
    if (!df.results$Sig_at_0.05.expression_in_week1[i]) {
        flag_at_0.05 <- paste0(flag_at_0.05, "0")
    } else if (df.results$logFC.expression_in_week1[i] < 0) {
        flag_at_0.05 <- paste0(flag_at_0.05, "-")
    } else if (df.results$logFC.expression_in_week1[i] > 0) {
        flag_at_0.05 <- paste0(flag_at_0.05, "+")
    }
    
    if (!df.results$Sig_at_0.1.expression_in_week1[i]) {
        flag_at_0.1 <- paste0(flag_at_0.1, "0")
    } else if (df.results$logFC.expression_in_week1[i] < 0) {
        flag_at_0.1 <- paste0(flag_at_0.1, "-")
    } else if (df.results$logFC.expression_in_week1[i] > 0) {
        flag_at_0.1 <- paste0(flag_at_0.1, "+")
    }
    
    # Flag4: Expression DE (Ctrl v 4707) in Week2
     if (!df.results$Sig_at_0.05.expression_in_week2[i]) {
        flag_at_0.05 <- paste0(flag_at_0.05, "0")
    } else if (df.results$logFC.expression_in_week2[i] < 0) {
        flag_at_0.05 <- paste0(flag_at_0.05, "-")
    } else if (df.results$logFC.expression_in_week2[i] > 0) {
        flag_at_0.05 <- paste0(flag_at_0.05, "+")
    }
    
    if (!df.results$Sig_at_0.1.expression_in_week2[i]) {
        flag_at_0.1 <- paste0(flag_at_0.1, "0")
    } else if (df.results$logFC.expression_in_week2[i] < 0) {
        flag_at_0.1 <- paste0(flag_at_0.1, "-")
    } else if (df.results$logFC.expression_in_week2[i] > 0) {
        flag_at_0.1 <- paste0(flag_at_0.1, "+")
    }
    
    df.results$DE_flag_at_0.05[i] <- flag_at_0.05
    df.results$DE_flag_at_0.1[i] <- flag_at_0.1
}

rm(flag_at_0.05, flag_at_0.1)
```

```{r}
# categorize each gene based on the 4 differential expression relationships, using "flags" at each category
# Flag1: Differentiation DE (week1 v week2) in CONTROL
# Flag2: Differentiation DE (week1 v week2) in 4707
# Flag3: Expression DE (Ctrl v 4707) in Week1
# Flag4: Expression DE (Ctrl v 4707) in Week2

# Flags based on direction of logFC (LFC) and whether there is a significant difference
# -: negative LFC
# 0: non-significant LFC
# +: positive LFC

df.results %>%
    group_by(DE_flag_at_0.05) %>%
    summarise(count = n(),
              interaction_true = sum(Sig_at_0.05.interaction),
              mirdb_targets = sum(miRDB_predicted_target),
              targetscan_targets = sum(TargetScan_predicted_target),
              mirtarbase_targets = sum(miRTarBase_validated_target)) %>%
    arrange(-count) -> df.result.groups

kable(df.result.groups)

df.results %>%
    dplyr::filter(Sig_at_0.05.interaction) %>%
    group_by(DE_flag_at_0.05) %>%
    summarise(count = n(),
              mirdb_targets = sum(miRDB_predicted_target),
              targetscan_targets = sum(TargetScan_predicted_target),
              mirtarbase_targets = sum(miRTarBase_validated_target)) %>%
    arrange(-count) -> df.result.groups.int

kable(df.result.groups.int)

```

```{r}
rownames(corrected.expression) <- fData(filt.eSet)$SYMBOL

df.samples <- as_tibble(pData(filt.eSet))
df.samples %<>%
    mutate(Week = as.numeric(Timepoint))
```

### 0000

```{r}
df.results %>%
    filter(DE_flag_at_0.05 == "0000" & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))


```

### 0+0-

```{r}
df.results %>%
    filter(DE_flag_at_0.05 == "0+0-" & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[2]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))
```

### 0-0+

```{r}
df.results %>%
    filter(DE_flag_at_0.05 == "0-0+" & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[2]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[3]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))
```

### 0++0

```{r}
df.results %>%
    filter((DE_flag_at_0.05 == "0++0" | DE_flag_at_0.05 == "0+++") & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[40]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[30]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))
```

```{r}
# df.results %>%
#     dplyr::filter(DE_flag_at_0.05 == "0++0" & Sig_at_0.05.interaction) %>%
#     pull(SYMBOL) -> genes.to.test

genes.to.test <- genes

# topGo data frame file for this analysis
df.topgo.rds <- paste0(dir.topgo, "topgo_results_0++0_classic_fisher_df.rds")

if (file.exists(df.topgo.rds)) {
    df.topgo.classic.fisher <- readRDS(df.topgo.rds)
    rm(df.topgo.rds)
} else {
    
    geneList <- factor(as.integer(all.genes %in% genes.to.test))
    names(geneList) <- all.genes
    
    myGOdata <- new("topGOdata",
                    ontology = "BP",
                    description = "my data",
                    allGenes = geneList,
                    nodeSize = 10,
                    annotationFun = annFUN.org,
                    mapping = "org.Hs.eg.db",
                    ID = "symbol")
    
    resultsFisher <- runTest(myGOdata, algorithm = "classic", statistic = "fisher")
    
    
    df.topgo.classic.fisher <- GenTable(myGOdata, classic_fisher = resultsFisher, topNodes = 80, orderBy = "classic_fisher")
    
    saveRDS(df.topgo.classic.fisher, df.topgo.rds)
    
    rm(genes.to.test, geneList, myGOdata, resultsFisher, df.topgo.rds)
}

kable(df.topgo.classic.fisher[1:40,])
```

### 0--0

```{r}
df.results %>%
    filter((DE_flag_at_0.05 == "0--0" | DE_flag_at_0.05 == "0---") & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[20]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[30]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))
```

```{r}
# df.results %>%
#     dplyr::filter(DE_flag_at_0.05 == "0--0" & Sig_at_0.05.interaction) %>%
#     pull(SYMBOL) -> genes.to.test

genes.to.test <- genes

# topGo data frame file for this analysis
df.topgo.rds <- paste0(dir.topgo, "topgo_results_0--0_classic_fisher_df.rds")

if (file.exists(df.topgo.rds)) {
    df.topgo.classic.fisher <- readRDS(df.topgo.rds)
    rm(df.topgo.rds)
} else {
    
    geneList <- factor(as.integer(all.genes %in% genes.to.test))
    names(geneList) <- all.genes
    
    myGOdata <- new("topGOdata",
                    ontology = "BP",
                    description = "my data",
                    allGenes = geneList,
                    nodeSize = 10,
                    annotationFun = annFUN.org,
                    mapping = "org.Hs.eg.db",
                    ID = "symbol")
    
    resultsFisher <- runTest(myGOdata, algorithm = "classic", statistic = "fisher")
    
    
    df.topgo.classic.fisher <- GenTable(myGOdata, classic_fisher = resultsFisher, topNodes = 80, orderBy = "classic_fisher")
    
    saveRDS(df.topgo.classic.fisher, df.topgo.rds)
    
    rm(genes.to.test, geneList, myGOdata, resultsFisher, df.topgo.rds)
}

kable(df.topgo.classic.fisher[1:40,])
```

### +0-0

```{r}
df.results %>%
    filter((DE_flag_at_0.05 == "+0-0" | DE_flag_at_0.05 == "+0--") & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[10]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[30]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))
```

```{r}
# df.results %>%
#     dplyr::filter(DE_flag_at_0.05 == "+0-0" & Sig_at_0.05.interaction) %>%
#     pull(SYMBOL) -> genes.to.test

genes.to.test <- genes

# topGo data frame file for this analysis
df.topgo.rds <- paste0(dir.topgo, "topgo_results_+0-0_classic_fisher_df.rds")

if (file.exists(df.topgo.rds)) {
    df.topgo.classic.fisher <- readRDS(df.topgo.rds)
    rm(df.topgo.rds)
} else {
    
    geneList <- factor(as.integer(all.genes %in% genes.to.test))
    names(geneList) <- all.genes
    
    myGOdata <- new("topGOdata",
                    ontology = "BP",
                    description = "my data",
                    allGenes = geneList,
                    nodeSize = 10,
                    annotationFun = annFUN.org,
                    mapping = "org.Hs.eg.db",
                    ID = "symbol")
    
    resultsFisher <- runTest(myGOdata, algorithm = "classic", statistic = "fisher")
    
    
    df.topgo.classic.fisher <- GenTable(myGOdata, classic_fisher = resultsFisher, topNodes = 80, orderBy = "classic_fisher")
    
    saveRDS(df.topgo.classic.fisher, df.topgo.rds)
    
    rm(genes.to.test, geneList, myGOdata, resultsFisher, df.topgo.rds)
}

kable(df.topgo.classic.fisher[1:40,])
```

### -0+0

```{r}
df.results %>%
    filter((DE_flag_at_0.05 == "-0+0" | DE_flag_at_0.05 == "-0++") & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[10]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[12]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))
```

```{r}
# df.results %>%
#     dplyr::filter(DE_flag_at_0.05 == "-0+0" & Sig_at_0.05.interaction) %>%
#     pull(SYMBOL) -> genes.to.test

genes.to.test <- genes

# topGo data frame file for this analysis
df.topgo.rds <- paste0(dir.topgo, "topgo_results_-0+0_classic_fisher_df.rds")

if (file.exists(df.topgo.rds)) {
    df.topgo.classic.fisher <- readRDS(df.topgo.rds)
    rm(df.topgo.rds)
} else {
    
    geneList <- factor(as.integer(all.genes %in% genes.to.test))
    names(geneList) <- all.genes
    
    myGOdata <- new("topGOdata",
                    ontology = "BP",
                    description = "my data",
                    allGenes = geneList,
                    nodeSize = 10,
                    annotationFun = annFUN.org,
                    mapping = "org.Hs.eg.db",
                    ID = "symbol")
    
    resultsFisher <- runTest(myGOdata, algorithm = "classic", statistic = "fisher")
    
    
    df.topgo.classic.fisher <- GenTable(myGOdata, classic_fisher = resultsFisher, topNodes = 80, orderBy = "classic_fisher")
    
    saveRDS(df.topgo.classic.fisher, df.topgo.rds)
    
    rm(genes.to.test, geneList, myGOdata, resultsFisher, df.topgo.rds)
}

kable(df.topgo.classic.fisher[1:40,])
```


### 0+++

```{r}
df.results %>%
    filter(DE_flag_at_0.05 == "0+++" & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[5]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[10]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))
```

```{r}
df.results %>%
    dplyr::filter(DE_flag_at_0.05 == "0+++" & Sig_at_0.05.interaction) %>%
    pull(SYMBOL) -> genes.to.test

#genes.to.test

# topGo data frame file for this analysis
df.topgo.rds <- paste0(dir.topgo, "topgo_results_0+++_classic_fisher_df.rds")

if (file.exists(df.topgo.rds)) {
    df.topgo.classic.fisher <- readRDS(df.topgo.rds)
    rm(df.topgo.rds)
} else {
    
    geneList <- factor(as.integer(all.genes %in% genes.to.test))
    names(geneList) <- all.genes
    
    myGOdata <- new("topGOdata",
                    ontology = "BP",
                    description = "my data",
                    allGenes = geneList,
                    nodeSize = 10,
                    annotationFun = annFUN.org,
                    mapping = "org.Hs.eg.db",
                    ID = "symbol")
    
    resultsFisher <- runTest(myGOdata, algorithm = "classic", statistic = "fisher")
    
    
    df.topgo.classic.fisher <- GenTable(myGOdata, classic_fisher = resultsFisher, topNodes = 80, orderBy = "classic_fisher")
    
    saveRDS(df.topgo.classic.fisher, df.topgo.rds)
    
    rm(genes.to.test, geneList, myGOdata, resultsFisher, df.topgo.rds)
}

kable(df.topgo.classic.fisher[1:40,])
```


```{r}
genes <- c("TSPAN6", "GRIA1", "GRIA2")

genes

gene <- genes[1]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[2]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))

gene <- genes[3]

tmp <- as.data.frame(corrected.expression[gene,])
tmp$CEL <- rownames(tmp)
colnames(tmp) <- c("expression", "CEL")

df.samples %>%
    left_join(tmp, by = "CEL") %>%
    ggplot(aes(x = Week, y = expression, color = Expression)) +
    geom_point() +
    geom_smooth(method = "lm") +
    labs(y = paste(gene, "expression"))
```
